### PR TITLE
fix(amplify_auth_cognito): handles new hosted ui android activity

### PR DESCRIPTION
**Pull request by haverchuck**
*Issue #, if available:*
#677 

*Description of changes:*
Amplify-Android has added a new activity that customers can use with an intent filter and will handle OAuth response on behalf of customers. However, our existing code has its own OAuth response handler which will periodically throw invalid grant exceptions if the customer uses the new activity.  This PR conditionally calls the response handler only if the customer is not using the new activity.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AuthCognito.kt [hunk 1]
```diff
@@ -119,12 +120,13 @@ public class AuthCognito : FlutterPlugin, ActivityAware, MethodCallHandler, Plug
...
-      }  else {
-        return false
-      }
+    var isHostedUIActivity = isRedirectActivityDeclared(context)
+    if (!isHostedUIActivity && requestCode == AWSCognitoAuthPlugin.WEB_UI_SIGN_IN_ACTIVITY_CODE) {
+      Amplify.Auth.handleWebUISignInResponse(data)
+      return true
+    }  else {
+      return false
+    }
```
Comment (cshfang) [inline packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AuthCognito.kt:129]: Nit: Suggested minor improvement on existing code
```suggestion
    }
    return false
```

File: packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/utils/IsRedirectActivityDeclared.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package com.amazonaws.amplify.amplify_auth_cognito.utils
```
Comment (cshfang) [inline packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/utils/IsRedirectActivityDeclared.kt:1]: Should we add license?

File: packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/utils/IsRedirectActivityDeclared.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                    return true
+                }
+            }
+        }
+        LOG.warn("$REDIRECT_ACTIVITY_NAME is not declared in AndroidManifest.")
+    } catch (error: java.lang.Exception) {
+        LOG.warn("Failed to inspect packages.")
+    }
+    return false
+}
```
Comment (cshfang) [inline packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/utils/IsRedirectActivityDeclared.kt:45]: Nit: Newline

File: packages/amplify_auth_cognito/android/src/test/kotlin/com/amazonaws/amplify/amplify_auth_cognito/IsRedirectActivityDeclaredTest.kt [hunk 4]
```diff
@@ -0,0 +1,54 @@
+package com.amazonaws.amplify.amplify_auth_cognito
```
Comment (cshfang) [inline packages/amplify_auth_cognito/android/src/test/kotlin/com/amazonaws/amplify/amplify_auth_cognito/IsRedirectActivityDeclaredTest.kt:1]: Should we add license?

File: packages/amplify_auth_cognito/android/src/test/kotlin/com/amazonaws/amplify/amplify_auth_cognito/IsRedirectActivityDeclaredTest.kt [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+     */
+    @Test
+    fun test_isRedirectActivity_HostedUIRedirectActivity() {
+        activityInfo.name = "HostedUIRedirectActivity"
+        var isHostedUIRedirectActivity = isRedirectActivityDeclared(context)
+        assert(isHostedUIRedirectActivity)
+    }
+
+    /**
+    * Test that the isRedirectActivityDeclared method will return true
```
Comment (cshfang) [inline packages/amplify_auth_cognito/android/src/test/kotlin/com/amazonaws/amplify/amplify_auth_cognito/IsRedirectActivityDeclaredTest.kt:43]:
```suggestion
    * Test that the isRedirectActivityDeclared method will return false
```

File: packages/amplify_auth_cognito/android/src/test/kotlin/com/amazonaws/amplify/amplify_auth_cognito/IsRedirectActivityDeclaredTest.kt (hunk 4 above)
Comment (haverchuck) [reply packages/amplify_auth_cognito/android/src/test/kotlin/com/amazonaws/amplify/amplify_auth_cognito/IsRedirectActivityDeclaredTest.kt:1]: Good catch - thanks!

File: packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/utils/IsRedirectActivityDeclared.kt (hunk 2 above)
Comment (haverchuck) [reply packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/utils/IsRedirectActivityDeclared.kt:1]: Good catch - thanks!

File: packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AuthCognito.kt (hunk 1 above)
Comment (haverchuck) [reply packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AuthCognito.kt:129]: Fixed on my end.

File: packages/amplify_auth_cognito/android/src/test/kotlin/com/amazonaws/amplify/amplify_auth_cognito/IsRedirectActivityDeclaredTest.kt [hunk 6]
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (cshfang) [inline packages/amplify_auth_cognito/android/src/test/kotlin/com/amazonaws/amplify/amplify_auth_cognito/IsRedirectActivityDeclaredTest.kt:2]:
```suggestion
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

**Review by cshfang — APPROVED** [review]

**Review by HuiSF — APPROVED** [review]

**Review by cshfang — APPROVED** [review]

**Pull request merged**